File: client/src/api/sqlSource.ts
/**
 * SQL source de-identification API, saved connections, schema inspection, and
 * async table export.  Mirrors the backend endpoints under /v1:
 *
 *   GET    /v1/sql-connections
 *   POST   /v1/sql-connections
 *   POST   /v1/sql-connections/{id}/test
 *   DELETE /v1/sql-connections/{id}
 *   POST   /v1/process/sql/inspect
 *   POST   /v1/jobs/sql-export
 *
 * Connections are read-only and host allow-listed server-side; passwords are
 * encrypted at rest and never returned by the API.
 */

import { fetchApi } from "./client";

/** A table-scoped column rule authored in the SQL schema explorer. */
export interface SqlColumnRule {
  table: string;
  column: string;
  action: string;
  params?: Record<string, unknown>;
}

export interface SqlConnection {
  id: string;
  name: string;
  host: string;
  port: number;
  dbname: string;
  username: string;
  sslmode: string;
  created_at: string;
}

export interface SqlConnectionCreate {
  name: string;
  host: string;
  port: number;
  dbname: string;
  username: string;
  password: string;
  sslmode?: string;
}

/** A column within a reflected table, same explorer vocabulary as TabularColumn. */
export interface SqlColumn {
  name: string;
  data_type: string;
  samples: string[];
  recommended_action?: string;
}

export interface SqlTable {
  name: string;
  row_estimate: number;
  columns: SqlColumn[];
}

export interface SqlSchemaPreview {
  schema: string;
  tables: SqlTable[];
}

export async function listSqlConnections(): Promise<SqlConnection[]> {
  const { connections } = await fetchApi<{ connections: SqlConnection[] }>(
    "/v1/sql-connections",
  );
  return connections;
}

export async function createSqlConnection(
  body: SqlConnectionCreate,
): Promise<SqlConnection> {
  return fetchApi<SqlConnection>("/v1/sql-connections", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

export async function testSqlConnection(
  id: string,
): Promise<{ ok: boolean; server: string }> {
  return fetchApi<{ ok: boolean; server: string }>(
    `/v1/sql-connections/${encodeURIComponent(id)}/test`,
    { method: "POST" },
  );
}

export async function deleteSqlConnection(id: string): Promise<void> {
  await fetchApi<void>(`/v1/sql-connections/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
}

export async function inspectSql(
  connectionId: string,
  schema = "public",
): Promise<SqlSchemaPreview> {
  return fetchApi<SqlSchemaPreview>("/v1/process/sql/inspect", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ connection_id: connectionId, schema }),
  });
}

export type SqlOutputFormat = "csv" | "ndjson" | "parquet";

export interface SqlExportJob {
  job_id: string;
  status: string;
}

/**
 * Submit an async sql-export job. ``rules`` are table-scoped column rules
 * (serialized as match: "table:<t>/column:<c>") so a single job can map many
 * tables unambiguously; when omitted the saved ``configProfile`` is used.
 * Returns the job id (track + download the ZIP on the Jobs page).
 */
export async function submitSqlExport(opts: {
  connectionId: string;
  tables: string[];
  schema?: string;
  outputFormat: SqlOutputFormat;
  rules?: SqlColumnRule[];
  configProfile?: string;
}): Promise<SqlExportJob> {
  const body: Record<string, unknown> = {
    connection_id: opts.connectionId,
    tables: opts.tables,
    schema: opts.schema ?? "public",
    output_format: opts.outputFormat,
  };
  if (opts.rules && opts.rules.length > 0) {
    body.rules = opts.rules.map((r) => ({
      match: `table:${r.table}/column:${r.column}`,
      action: r.action,
      ...(r.params ? { params: r.params } : {}),
    }));
  } else if (opts.configProfile) {
    body.config_profile = opts.configProfile;
  }
  return fetchApi<SqlExportJob>("/v1/jobs/sql-export", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

File: services/anonymizer/src/api/routers/sql_source.py
"""SQL source de-identification endpoints.

Saved connections (read-only, host allow-listed, encrypted passwords) to external
PostgreSQL databases, schema inspection (the column explorer at DB scale), and an
async export job that de-identifies selected tables to files.

  POST   /v1/sql-connections            create + validate a connection (admin)
  GET    /v1/sql-connections            list connections (no passwords)
  POST   /v1/sql-connections/{id}/test  open+close a read-only connection
  DELETE /v1/sql-connections/{id}       remove a connection (admin)
  POST   /v1/process/sql/inspect        reflect schema → tables/columns + recommendations

The async ``sql-export`` job is submitted via ``POST /v1/jobs/sql-export``
(in ``api/routers/jobs.py``).
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, field_validator

from api.auth import AuthContext
from api.deps import limiter
from api.services.sql_source import (
    SqlConnectionNotFound,
    SqlSourceService,
    SqlStoreUnavailable,
)
from integrations.sql_source.connect import SqlSourceError
from integrations.sql_source.secrets import CredentialKeyError

router = APIRouter()
logger = logging.getLogger("medanon")

_service = SqlSourceService()


def _require_admin(request: Request) -> None:
    auth: AuthContext | None = getattr(request.state, "auth", None)
    if auth is None or not auth.has_role("admin"):
        raise HTTPException(
            status_code=403, detail="Admin role required to manage SQL connections."
        )


class SqlConnectionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(default=5432, ge=1, le=65535)
    dbname: str = Field(min_length=1, max_length=128)
    username: str = Field(min_length=1, max_length=128)
    password: str = Field(default="", max_length=512)
    sslmode: str = Field(default="prefer")

    @field_validator("sslmode")
    @classmethod
    def _valid_sslmode(cls, v: str) -> str:
        allowed = {"disable", "allow", "prefer", "require", "verify-ca", "verify-full"}
        if v not in allowed:
            raise ValueError(f"sslmode must be one of {sorted(allowed)}")
        return v


class SqlInspectRequest(BaseModel):
    connection_id: str
    schema_name: str = Field(default="public", max_length=128, alias="schema")

    model_config = {"populate_by_name": True}


def _map_error(exc: Exception) -> HTTPException:
    """Translate domain errors to clean HTTP responses (never leak internals)."""
    if isinstance(exc, SqlStoreUnavailable):
        return HTTPException(status_code=503, detail=str(exc))
    if isinstance(exc, SqlConnectionNotFound):
        return HTTPException(status_code=404, detail=f"Connection not found: {exc}")
    if isinstance(exc, (SqlSourceError, CredentialKeyError)):
        return HTTPException(status_code=422, detail=str(exc))
    # Unknown failure  log the real cause (type + message) so it is diagnosable
    # without leaking internals to the client.
    logger.error("sql_source_unexpected_error: %s: %s", type(exc).__name__, exc)
    return HTTPException(status_code=500, detail="SQL source error")


@router.post("/sql-connections", status_code=201)
@limiter.limit("30/minute")
async def create_sql_connection(body: SqlConnectionCreate, request: Request):
    """Save a new SQL source connection (password encrypted at rest). Admin only."""
    _require_admin(request)
    try:
        return _service.create_connection(
            name=body.name,
            host=body.host,
            port=body.port,
            dbname=body.dbname,
            username=body.username,
            password=body.password,
            sslmode=body.sslmode,
        )
    except Exception as exc:
        raise _map_error(exc) from exc


@router.get("/sql-connections")
async def list_sql_connections(request: Request):
    """List saved connections (never includes passwords)."""
    try:
        return {"connections": _service.list_connections()}
    except Exception as exc:
        raise _map_error(exc) from exc


@router.post("/sql-connections/{conn_id}/test")
@limiter.limit("30/minute")
async def test_sql_connection(conn_id: str, request: Request):
    """Open and immediately close a read-only connection to verify it works."""
    try:
        return await _service.test_connection(conn_id)
    except Exception as exc:
        raise _map_error(exc) from exc


@router.delete("/sql-connections/{conn_id}", status_code=204)
async def delete_sql_connection(conn_id: str, request: Request):
    """Delete a saved connection. Admin only."""
    _require_admin(request)
    try:
        if not _service.delete_connection(conn_id):
            raise HTTPException(status_code=404, detail="Connection not found")
    except HTTPException:
        raise
    except Exception as exc:
        raise _map_error(exc) from exc


@router.post("/process/sql/inspect")
@limiter.limit("30/minute")
async def inspect_sql(body: SqlInspectRequest, request: Request):
    """Reflect a connection's schema for the column explorer (no data leaves as-is).

    Returns ``{schema, tables: [{name, row_estimate, columns: [{name, data_type,
    samples, recommended_action}]}]}``.
    """
    try:
        return await _service.inspect(body.connection_id, body.schema_name)
    except Exception as exc:
        raise _map_error(exc) from exc
